pkg/apis/messaging/v1beta1/channel_defaults.go

```diff
@@ -40,10 +42,26 @@ func defaultSecretSelector() *corev1.SecretKeySelector {
 }
 
 func (c *Channel) SetDefaults(ctx context.Context) {
+	// We need to set this to the _stored_ version of the Channel. If we set to anything other than
+	// the stored version, then when reading the stored version, conversion won't be called so
+	// nothing will set it to the stored version.
+	// Note that if a user sends a bad version of this annotation (e.g. sets it to v1beta1), then we
+	// won't overwrite their bad input. This is because the webhook:
+	// 1. Reads the stored version.
+	// 2. Converts to the desired version.
+	// 3. Defaults the desired version.
+	// So we don't know if the user or the converter put the value here, therefore we are forced to
+	// assume it was the converter and shouldn't change it.
```

Comment:
Do you mean the conversion happens before this function is invoked so that we have to default to stored version?

Reply:
Yes, conversion occurs before this function is invoked. 

This actually isn't needed in the v1beta1 code path. It is only required in the storage version path (v1alpha1). It is only needed for the storage version, because we normally set this value during conversion, but if the storage version is requested, then no conversion occurs, so nothing sets it. I chose to add it here too so that we would not forget it when switching the storage version.

Eventing has the same logic, although only in the non-storage path:

https://github.com/knative/eventing/blob/d76621b63ac486411fdac4b9da07fd3a7a8e45b1/pkg/apis/messaging/v1beta1/in_memory_channel_defaults.go#L26-L36

Reply:
Thanks for explaining!